include/envoy/network/listen_socket.h

```diff
@@ -30,36 +31,41 @@ class Socket {
    */
   virtual void close() PURE;
 
+  enum class SocketState { PreBind, PostBind };
+
   /**
    * Visitor class for setting socket options.
    */
-  class Options {
+  class Option {
   public:
-    virtual ~Options() {}
+    virtual ~Option() {}
 
     /**
      * @param socket the socket on which to apply options.
+     * @param state the current state of the socket. Significant for options that can only be
+     *        set for some particular state of the socket.
      * @return true if succeeded, false otherwise.
      */
-    virtual bool setOptions(Socket& socket) const PURE;
+    virtual bool setOption(Socket& socket, SocketState state) const PURE;
 
     /**
-     * @return bits that can be used to separate connections based on the options. Should return
-     *         zero if connections with different options can be pooled together. This is limited
-     *         to 32 bits to allow these bits to be efficiently combined into a larger hash key
-     *         used in connection pool lookups.
+     * @param vector of bits that can be used to separate connections based on the options. Should
+     *        return zero if connections with different options can be pooled together. This is
+     *        limited to 32 bits to allow these bits to be efficiently combined into a larger hash
+     *        key used in connection pool lookups.
      */
-    virtual uint32_t hashKey() const PURE;
+    virtual void hashKey(std::vector<uint8_t>& key) const PURE;
   };
-  typedef std::shared_ptr<Options> OptionsSharedPtr;
+  typedef std::shared_ptr<Option> OptionSharedPtr;
+  typedef std::shared_ptr<std::vector<OptionSharedPtr>> OptionsSharedPtr;
```

Comment:
In general, shared_ptr of vector of shared_ptr doesn't sound like a good practice, is it possible to make either to unique_ptr with clearer ownership?

Reply:
I'll make it a shared vector of unique option pointers, transferring the ownership of the options to the socket, but allowing the whole set be shared.